[DOCS] Update documentation for index sorting and routing for logsdb (#120721)
elastic/elasticsearch: docs/reference/data-streams/logs.asciidoc

[[logs-data-stream]]
== Logs data stream

IMPORTANT: The {es} `logsdb` index mode is generally available in Elastic Cloud Hosted 
and self-managed Elasticsearch as of version 8.17, and is enabled by default for 

> logs in https://www.elastic.co/elasticsearch/serverless[{serverless-full}]. 

Author: IntelliJ took the liberty, for some reason..

=== Synthetic source

If you have the required https://www.elastic.co/subscriptions[subscription], `logsdb` index mode uses <<synthetic-source,synthetic `_source`>>, which omits storing the original `_source`
field. Instead, the document source is synthesized from doc values or stored fields upon document retrieval.

If you don't have the required https://www.elastic.co/subscriptions[subscription], `logsdb` mode uses the original `_source` field.

Before using synthetic source, make sure to review the <<synthetic-source-restrictions,restrictions>>.

When working with multi-value fields, the `index.mapping.synthetic_source_keep` setting controls how field values
are preserved for <<synthetic-source,synthetic source>> reconstruction. In `logsdb`, the default value is `arrays`,
which retains both duplicate values and the order of entries. However, the exact structure of
array elements and objects is not necessarily retained. Preserving duplicates and ordering can be critical for some
log fields, such as DNS A records, HTTP headers, and log entries that represent sequential or repeated events.

[discrete]
[[logsdb-sort-settings]]
=== Index sort settings

> In `logsdb` index mode, indices are sorted by fields `host.name` and `@timestamp` by default. The `@timestamp` field is

Reviewer: These additions (starting with this line) are a bit dense and make the page less skimmable, so my suggestions break up the text for clarity and readability.

automatically injected if it is not present. The `host.name` field is automatically injected as `keyword` if it is not
present and can be injected - this may not be possible if  `host` is a keyword field, for instance. If field
`host.name` can't be injected or can't be used for sorting (e.g. it's an IP field), sorting is only applied to field
`@timestamp`.

`index.sort.field`: `["host.name", "@timestamp"]`::
Indices are sorted by `host.name` and `@timestamp` by default. The `@timestamp` field is automatically injected if it is not present.
NOTE: If `host.name` is injected and `subobjects` is set to `true` (default), the `host` field is mapped as an object
field named `host` with a `name` child field of type `keyword`. If `subobjects` is set to `false`, a single
`host.name` field is mapped as a `keyword` field.

`index.sort.order`: `["desc", "desc"]`::
Both `host.name` and `@timestamp` are sorted in descending (`desc`) order, prioritizing the latest data.
`host.name` and `@timestamp` are sorted in ascending and descending order respectively, prioritizing the latest data.

`index.sort.mode`: `["min", "min"]`::

> The `min` mode sorts indices by the minimum value of multi-value fields.

Author: The logic in the code has changed, so these comments no longer hold - hence I updated the text.

> The `min` mode sorts indices by the minimum value of multi-value fields.

Reviewer: thanks! resolving

> automatically added to the list of sort fields. It is highly recommended to include it manually, as the last sort
> field with `desc` ordering.

Reviewer: ```suggestion
automatically added to the list of sort fields. For best results, include it manually as the last sort
field, with `desc` ordering.
```

* **Switch to a different <<index-mode-setting,index mode>>**: If resolving `host.name` mapping conflicts is not feasible, you can choose not to use `logsdb` mode.

IMPORTANT: On existing data streams, `logsdb` mode is applied on <<data-streams-rollover,rollover>> (automatic or manual).

[discrete]
[[logsdb-sort-routign]]
==== Optimized routing on sort fields

> The storage footprint of `logsdb` indexes can further be reduced by enabling a routing optimization that relies on
> the fields in the sort configuration (except for `@timestamp`) to route documents to shards. The storage wins depend on
> the sort configuration and the nature of the logged data - we observed 20% storage reductions in our benchmarks,
> compared to the default configuration for `logsdb` mode. Combined with a negligible penalty to ingest
> performance (1-4%), this optimization is a good option for data streams that are expected to grow substantially with
> time.

Reviewer: ```suggestion
To reduce the storage footprint of `logsdb` indexes, you can enable routing optimizations. A routing optimization uses the fields in the sort configuration (except for `@timestamp`) to route documents to shards. 

In benchmarks,
routing optimizations reduced storage requirements by 20% compared to the default `logsdb` configuration, with a negligible penalty to ingestion
performance (1-4%). Routing optimizations can benefit data streams that are expected to grow substantially over
time. Exact results depend on the sort configuration and the nature of the logged data.
```

Reviewer: cc @tylerperk: Are you OK with the additions made in this PR, especially the benchmarking and performance stats? Not sure whether they might be a maintenance headache, especially if there are better sources like https://elasticsearch-benchmarks.elastic.co/

(with or without my edits -- just wondering about the additions in general)

> Configuring the routing optimization requires the following:

Reviewer: ```suggestion
To configure a routing optimization:
```

>  * Include index setting `[index.logsdb.route_on_sort_fields:true]` in the data stream configuration.
>  * <<index-modules-index-sorting, Configure index sorting>> with 2 or more fields, in addition to `@timestamp`.
>  * Make sure <<mapping-id-field, field `_id`>> is not populated in ingested documents, as it needs to get
>    auto-generated.

Reviewer: ```suggestion
 * Include the index setting `[index.logsdb.route_on_sort_fields:true]` in the data stream configuration.
 * <<index-modules-index-sorting, Configure index sorting>> with two or more fields, in addition to `@timestamp`.
 * Make sure the <<mapping-id-field,`_id`>> field is not populated in ingested documents. It should be
   auto-generated instead.
```

> Using a custom sort configuration is required to minimize the possibility of creating hotspots, in case of a
> logging spike producing documents that all get routed to a single shard. To prevent this, and to improve storage
> efficiency, it is recommended to use a few fields that have a rather low cardinality and don't co-vary
> (e.g. `host.name` and `host.id` are likely a bad choice).

Reviewer: ```suggestion
Logging spikes can cause hotspots by producing documents that all get routed to a single 
shard. To prevent hotspots and improve storage efficiency, your configuration should use a few sort fields that have a relatively low cardinality and don't co-vary (for example, `host.name` and `host.id` are not optimal).
```

Author: The logic requires a custom sort config to reduce the likelihood of hotspots, as opposed to working with the default sort config. I think the updated text (and my version, possibly) missed this part. Maybe we can clarify this better?

Reviewer: Ah, thanks, I see what you mean. I'd suggest losing the logging spikes sentence -- WDYT of this?

```suggestion
A custom sort configuration is required, to minimize hotspots and improve storage efficiency. For best results, use a few sort fields that have a relatively low cardinality and don't co-vary
(for example, `host.name` and `host.id` are not optimal).
```

Author: Sounds good, though it'd be nice to explain what leads to hotspots - I don't think this is mentioned elsewhere in this page. Another possibility is to include such a note above, where we describe the option for custom sort config.

Reviewer: OK one more try 🙂 

```
A custom sort configuration is required, to improve storage efficiency and to
minimize hotspots from logging spikes that route documents to a single shard. 
For best results, use a few sort fields that have a relatively low cardinality and
don't co-vary (for example, `host.name` and `host.id` are not optimal).
```
